c10/core/ScalarType.h

```diff
@@ -297,10 +297,40 @@ static inline ScalarType toUnderlying(ScalarType t) {
   }
 }
 
+static inline bool isSignedType(ScalarType t) {
+  if (t == ScalarType::Half) {
+    return true;
+  }
+  #define CASE_SIGNED(ctype, name) \
+    case ScalarType::name:                       \
+      return std::is_signed<ctype>();
+
+    switch (toUnderlying(t)) {
+      AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS(CASE_SIGNED)
+      default:
+        AT_ERROR("Unknown ScalarType");
+    }
+  #undef CASE_SIGNED
+}
+
 static inline bool isUnderlying(ScalarType type, ScalarType qtype) {
   return type == toUnderlying(qtype);
 }
 
+static inline bool canCast(const ScalarType from, const ScalarType to) {
+  // We disallow float -> integral, e.g., int_tensor *= float is disallowed.
+  if (isFloatingType(from) && isIntegralType(to, false)) {
+    return false;
+  }
+
+  // treat bool as a distinct category. This resolves most issues that arrise from
```

Comment:
nit: arise

Reply:
I kind of know what this comment means now, but I certainly won't know what it means in a few months.

Also, why is uint8 / int8 not symmetrically castable?